docs/proposals/config_data.md

```diff
@@ -0,0 +1,289 @@
+# Generic Configuration Object
+
+## Abstract
+
+This proposal proposes a generic configuration object named ``ConfigData`` stored inside ``etcd`` that
+stores data used for the configuration of applications deployed on ``Kubernetes``.
+
+The main focus points of this proposal are:
+
+* Dynamic distribution of configuration data to components and deployed applications.
+* Encapsulate configuration information and simplify ``Kubernetes`` deployments.
+* Create a flexible configuration model for ``Kubernetes``.
+
+## Motivation
+
+Currently command-line flags and environment variables are used for configurations in ``Kubernetes``
+which is complicated and error-prone. This approach is not suitable for a distributed computing environment.
+Specifically the problems with the current approach are:
+
+* Complicated deployment orchestration
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+
+## Use Cases
+
+1. Be able to load API resources as environment variables into a component.
+2. Be able to load API resources as mounted volumes into a component.
+3. Be able to update mounted volumes after a change to the API resource.
+
+## Solution
+
+A new ``ConfigData`` API resource is proposed to address these problems.
+
+The ``ConfigData`` architecture will be very similar to that of ``Secrets``.
+
+``ConfigData`` can either be mounted as a volume or can be injected as environment
+variables to a container through the use of the downward API.
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container.
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the API server every time the object is modified.  After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsibility to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object. This is especially true if ``ConfigData`` is injected as environment variables. Dynamic configuration with environment variables is out of scope).
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* Persistent configuration information through API versioning.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
```

Comment:
Remove this bullet or clarify what this means, remember that we do not want the proposal overly focused on Kubernetes components.